src/datatype/aggregate_svecf32.rs

```diff
@@ -0,0 +1,428 @@
+            self.values
+                .copy_within(((p + 1) as usize)..capacity, (i as usize) + 1);
+            ((capacity as i32) - p + i) as u32
+        };
+    }
+}
+
+/// accumulate intermediate state for sparse vector
+#[pgrx::pg_extern(sql = "\
+CREATE FUNCTION _vectors_svecf32_aggregate_avg_sum_sfunc(internal, svector) RETURNS internal IMMUTABLE PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '@FUNCTION_NAME@';")]
+fn _vectors_svecf32_aggregate_avg_sum_sfunc(
+    current: Internal,
+    value: Option<SVecf32Input<'_>>,
+    fcinfo: pgrx::pg_sys::FunctionCallInfo,
+) -> Internal {
+    let Some(value) = value else { return current };
+    let old_context = unsafe {
+        let mut agg_context: *mut ::pgrx::pg_sys::MemoryContextData = std::ptr::null_mut();
+        if ::pgrx::pg_sys::AggCheckCallContext(fcinfo, &mut agg_context) == 0 {
+            ::pgrx::error!("aggregate function called in non-aggregate context");
+        }
+        // pg will switch to a temporary memory context when call aggregate trans function.
+        // https://github.com/postgres/postgres/blob/52b49b796cc7fd976f4da6aa49c9679ecdae8bd5/src/backend/executor/nodeAgg.c#L761-L801
+        // If want to reuse the state in aggregate, we need to switch to aggregate context like https://github.com/postgres/postgres/blob/7c655a04a2dc84b59ed6dce97bd38b79e734ecca/src/backend/utils/adt/numeric.c#L5635-L5648.
+        ::pgrx::pg_sys::MemoryContextSwitchTo(agg_context)
+    };
+    let result = match unsafe { current.get_mut::<SVecf32AggregateAvgSumStype>() } {
+        // if the state is empty, copy the input vector
+        None => {
+            let internal = Internal::new(SVecf32AggregateAvgSumStype::new_with_capacity(
+                value.dims() as u32,
+                value.len(),
+            ));
+            let state = unsafe { internal.get_mut::<SVecf32AggregateAvgSumStype>().unwrap() };
+            state.merge_in_place(value.for_borrow());
+            state.count = 1;
+            internal
+        }
+        Some(state) => {
+            let dims = state.dims();
+            let value_dims = value.dims();
+            check_matched_dims(dims, value_dims);
+            let count = state.count() + 1;
+            if state.check_capacity(value.len()) {
+                // merge the input vector into state
+                state.merge_in_place(value.for_borrow());
+                state.count = count;
+                // return old state
+                current
+            } else {
+                // allocate a new state and merge the old state
+                let mut new_state = SVecf32AggregateAvgSumStype::new_with_capacity(
+                    dims as u32,
+                    state.len() + value.len(),
+                );
+                new_state.merge_in_place(SVecf32Borrowed::new(
+                    dims as u32,
+                    state.indexes(),
+                    state.values(),
+                ));
```

Comment:
`state` may not be a `SVecf32` because of zeros. `merge_in_place` should not accept `SVecf32Borrowed`.

Reply:
I think zero number will be filter in Line115.

Reply:
So what `filter_zero` is used for?

Reply:
Mainly for `avg` aggregate. Elements in `state` may be get zero after division by `count`.